src/common/dynarray.h

```diff
@@ -0,0 +1,335 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <algorithm>
+#include <initializer_list>
+#include <iostream>
+#include <iterator>
+#include <limits>
+#include <memory>
+#include <stdexcept>
+#include <tuple>
+#include <type_traits>
+#include <utility>
+
+#include "include/uses_allocator.h"
+
+/// The problem:
+///     - Non-movable types
+///     - Non-default-constructible types
+///     - Entries whose constructor parameters depend on the index
+
+/// The limitations:
+///     - No resize
+///     - No insert
+///     - No delete
+
+
+/// The solution, a class with all the methods of vector that are
+/// compatible with never moving an element (i.e. no reallocations)
+
+namespace ceph {
+template<typename T, typename Allocator = std::allocator<T>>
+class dynarray {
```

Comment:
have you tried using `ceph::containers::tiny_vector` here? it was added for a similar purpose in https://github.com/ceph/ceph/pull/28987